From 6dd307f44554f38d9d51ffedf155ca104c0f100e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20P=C3=ADsa=C5=99?= <ppisar@redhat.com>
Date: Tue, 24 Jan 2023 16:55:12 +0100
Subject: [PATCH] Adapt to GCC 13
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

GCC 13 errors on implicit function declarations. This patch fixes.

It also corrects warnings about a mismatch between struct iptc_handle*
and struct xtc_handle* arguments. It does so by aliging
include/libiptc/libiptc.h to <libiptc/libiptc.h> as found in
iptables-1.8.9. I did not check how far this change is backward
compatible.

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 include/iptables.h        | 7 +++++++
 include/libiptc/libiptc.h | 9 ++++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/include/iptables.h b/include/iptables.h
index 84211c3..b6ea15a 100644
--- a/include/iptables.h
+++ b/include/iptables.h
@@ -4,6 +4,7 @@
 #include <netinet/ip.h>
 #include <xtables.h>
 #include <libiptc/libiptc.h>
+#include <libiptc/xtcshared.h>
 #include <iptables/internal.h>
 
 /* Your shared library should call one of these. */
@@ -17,6 +18,12 @@ extern int for_each_chain(int (*fn)(const ipt_chainlabel, int, struct iptc_handl
 		int verbose, int builtinstoo, struct iptc_handle *handle);
 extern void print_rule(const struct ipt_entry *e,
 		struct iptc_handle *handle, const char *chain, int counters);
+extern int delete_chain4(const xt_chainlabel chain, int verbose,
+		struct xtc_handle *handle);
+#if XTABLES_VERSION_CODE >= 7
+extern int do_command4(int argc, char *argv[], char **table,
+		struct xtc_handle **handle, bool restore);
+#endif
 
 /* kernel revision handling */
 extern int kernel_version;
diff --git a/include/libiptc/libiptc.h b/include/libiptc/libiptc.h
index 5d782da..3ed6c96 100644
--- a/include/libiptc/libiptc.h
+++ b/include/libiptc/libiptc.h
@@ -24,7 +24,8 @@ extern "C" {
 
 #define IPT_ALIGN(s) (((s) + ((IPT_MIN_ALIGN)-1)) & ~((IPT_MIN_ALIGN)-1))
 
-struct iptc_handle;
+struct xtc_handle;
+#define iptc_handle xtc_handle
 
 typedef char ipt_chainlabel[32];
 
@@ -88,6 +89,12 @@ int iptc_append_entry(const ipt_chainlabel chain,
 		      const struct ipt_entry *e,
 		      struct iptc_handle *handle);
 
+/* Check whether a mathching rule exists */
+int iptc_check_entry(const ipt_chainlabel chain,
+		     const struct ipt_entry *origfw,
+		     unsigned char *matchmask,
+		     struct iptc_handle *handle);
+
 /* Delete the first rule in `chain' which matches `e', subject to
    matchmask (array of length == origfw) */
 int iptc_delete_entry(const ipt_chainlabel chain,
-- 
2.39.1

